src/webgpu/api/validation/encoding/cmds/render/dynamic_state.spec.ts

```diff
@@ -130,71 +130,97 @@ class F extends ValidationTest {
+g.test('setViewport,width_height_nonnegative')
   .desc(
-    `Test that the parameters of setViewport to define the box must be non-negative.
+    `Test that the width and height parameters of setViewport must be non-negative.
 
 TODO Test -0 (it should be valid) but can't be tested because the harness complains about duplicate parameters.
 TODO Test the first value smaller than -0`
   )
   .paramsSubcasesOnly([
     // Control case: everything to 0 is ok, covers the empty viewport case.
-    { x: 0, y: 0, w: 0, h: 0 },
+    { w: 0, h: 0 },
 
     // Test -1
-    { x: -1, y: 0, w: 0, h: 0 },
-    { x: 0, y: -1, w: 0, h: 0 },
-    { x: 0, y: 0, w: -1, h: 0 },
-    { x: 0, y: 0, w: 0, h: -1 },
+    { w: -1, h: 0 },
+    { w: 0, h: -1 },
   ])
   .fn(t => {
-    const { x, y, w, h } = t.params;
-    const success = x >= 0 && y >= 0 && w >= 0 && h >= 0;
-    t.testViewportCall(success, { x, y, w, h, minDepth: 0, maxDepth: 1 });
+    const { w, h } = t.params;
+    const success = w >= 0 && h >= 0;
+    t.testViewportCall(success, { x: 0, y: 0, w, h, minDepth: 0, maxDepth: 1 });
   });
 
-g.test('setViewport,xy_rect_contained_in_attachment')
+g.test('setViewport,exceeds_attachment_size')
+  .desc(`Test that the viewport can exceed the attachment size`)
+  .paramsSubcasesOnly([
+    { attachmentWidth: 3, attachmentHeight: 3 },
+    { attachmentWidth: 1024, attachmentHeight: 1024 },
+  ])
+  .fn(t => {
+    const { attachmentWidth, attachmentHeight } = t.params;
+    t.testViewportCall(
+      true,
+      { x: 0, y: 0, w: attachmentWidth + 1, h: attachmentHeight + 1, minDepth: 0, maxDepth: 1 },
+      { width: attachmentWidth, height: attachmentHeight, depthOrArrayLayers: 1 }
+    );
+  });
+
+g.test('setViewport,xy_rect_contained_in_bounds')
   .desc(
-    'Test that the rectangle defined by x, y, width, height must be contained in the attachments'
+    `Test that the rectangle defined by x, y, width, height must be contained in the maximum viewport bounds
+and that the viewport size cannot exceed the maximum.`
   )
   .paramsSubcasesOnly(u =>
-    u
-      .combineWithParams([
-        { attachmentWidth: 3, attachmentHeight: 5 },
-        { attachmentWidth: 5, attachmentHeight: 3 },
-        { attachmentWidth: 1024, attachmentHeight: 1 },
-        { attachmentWidth: 1, attachmentHeight: 1024 },
-      ])
-      .combineWithParams([
-        // Control case: a full viewport is valid.
-        { dx: 0, dy: 0, dw: 0, dh: 0 },
-
-        // Other valid cases with a partial viewport.
-        { dx: 1, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: -1 },
-        { dx: 0, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: -1 },
-
-        // Test with a small value that causes the viewport to go outside the attachment.
-        { dx: 1, dy: 0, dw: 0, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: 0 },
-        { dx: 0, dy: 0, dw: 1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: 1 },
-      ])
+    u.combineWithParams([
```

Comment:
nit: maybe parameterize over the dimension so you don't have to list cases for x and y separately:
```ts
u
  .combine('dim', [0, 1])
  .combineWithParams([
    { m: 0, d: 0 },
    // ...
  ])
```

```ts
const xy = [0, 0];
xy[dim] = maxViewportSize * m + d;
```

Reply:
Nice idea! Done!